File: pytorch/cayley/sig.py
from collections.abc import Iterator

import torch
from torch import Tensor


def _popcount(n: int) -> int:
    c = 0
    while n:
        n &= n - 1
        c += 1
    return c


def _swap_sign(a: int, b: int, dim: int) -> int:
    # Parity of swaps to reorder the concatenation of the two basis blades
    # (as bit-listed sequences) into canonical low-bit-first order.
    count = 0
    for j in range(dim):
        above = a >> (j + 1)
        if (b >> j) & 1:
            count += _popcount(above & ((1 << (dim - j - 1)) - 1))
    return 1 if count % 2 == 0 else -1


def _basis_sign(i: int, p: int, q: int) -> int:
    if i < p:
        return 1
    if i < p + q:
        return -1
    return 0  # null direction, squares to zero


def _metric_sign(mask: int, p: int, q: int, dim: int) -> int:
    # Product of basis-vector signs for every set bit of `mask`; zero if any
    # bit lands in the null block.
    acc = 1
    for i in range(dim):
        if (mask >> i) & 1:
            m = _basis_sign(i, p, q)
            if m == 0:
                return 0
            acc *= m
    return acc


def _cayley_entries(p: int, q: int, r: int) -> Iterator[tuple[int, int, int, int]]:
    # Yields (ia, ib, ic, sign) for every nonzero entry of the Cl(p,q,r)
    # geometric-product Cayley tensor. Iterates the n^2 blade-mask pairs and
    # skips entries that fall into the null block.
    dim = p + q + r
    n = 1 << dim
    for a in range(n):
        for b in range(n):
            m = _metric_sign(a & b, p, q, dim)
            if m == 0:
                continue
            yield a, b, a ^ b, _swap_sign(a, b, dim) * m


def sparse_cayley_from_sig(
    p: int,
    q: int,
    r: int = 0,
    *,
    dtype: torch.dtype = torch.float32,
) -> tuple[Tensor, Tensor, Tensor, Tensor]:
    # Direct combinatorial construction of the sparse Cayley table for Cl(p,q,r).
    # Emits (ia, ib, ic, sign) without materialising the dense (n,n,n) tensor
    # first; at N=8 this avoids a 256 MB float32 allocation.
    ia_l: list[int] = []
    ib_l: list[int] = []
    ic_l: list[int] = []
    sgn_l: list[int] = []
    for a, b, c, s in _cayley_entries(p, q, r):
        ia_l.append(a)
        ib_l.append(b)
        ic_l.append(c)
        sgn_l.append(s)
    return (
        torch.tensor(ia_l, dtype=torch.int32),
        torch.tensor(ib_l, dtype=torch.int32),
        torch.tensor(ic_l, dtype=torch.int32),
        torch.tensor(sgn_l, dtype=dtype),
    )


def dense_cayley_from_sig(
    p: int,
    q: int,
    r: int = 0,
    *,
    dtype: torch.dtype = torch.float32,
) -> Tensor:
    # Dense (n,n,n) Cayley tensor for Cl(p,q,r). Useful as a reference for the
    # einsum baseline; prefer sparse_cayley_from_sig when only the kernel
    # arrays are needed.
    n = 1 << (p + q + r)
    C = torch.zeros(n, n, n, dtype=dtype)
    for a, b, c, s in _cayley_entries(p, q, r):
        C[a, b, c] = s
    return C

File: rust/src/sig.rs
pub struct Cayley {
    pub ia: Vec<i32>,
    pub ib: Vec<i32>,
    pub ic: Vec<i32>,
    pub sign: Vec<f32>,
}

fn popcount(mut n: u32) -> u32 {
    let mut c = 0;
    while n != 0 {
        n &= n - 1;
        c += 1;
    }
    c
}

fn swap_sign(a: u32, b: u32, dim: u32) -> i32 {
    let mut count = 0u32;
    for j in 0..dim {
        let above = a >> (j + 1);
        if (b >> j) & 1 == 1 {
            let mask = (1u32 << (dim - j - 1)) - 1;
            count += popcount(above & mask);
        }
    }
    if count.is_multiple_of(2) { 1 } else { -1 }
}

fn basis_sign(i: u32, p: u32, q: u32) -> i32 {
    if i < p {
        1
    } else if i < p + q {
        -1
    } else {
        0
    }
}

fn metric_sign(mask: u32, p: u32, q: u32, dim: u32) -> i32 {
    let mut acc = 1i32;
    for i in 0..dim {
        if (mask >> i) & 1 == 1 {
            let m = basis_sign(i, p, q);
            if m == 0 {
                return 0;
            }
            acc *= m;
        }
    }
    acc
}

pub fn sparse_cayley_from_sig(p: u32, q: u32, r: u32) -> Cayley {
    let dim = p + q + r;
    let n = 1u32 << dim;
    let mut ia = Vec::new();
    let mut ib = Vec::new();
    let mut ic = Vec::new();
    let mut sign = Vec::new();
    for a in 0..n {
        for b in 0..n {
            let m = metric_sign(a & b, p, q, dim);
            if m == 0 {
                continue;
            }
            let s = swap_sign(a, b, dim) * m;
            ia.push(a as i32);
            ib.push(b as i32);
            ic.push((a ^ b) as i32);
            sign.push(s as f32);
        }
    }
    Cayley { ia, ib, ic, sign }
}
